### PR TITLE
[SYCL][ESIMD] Add ESIMD tests on TPM

### PR DESCRIPTION
**Pull request by fveselov**

### CODE REVIEW

**Comment by fveselov** [comment]
Results of PR#2856-No.891 job

16:24:44  PASS: SYCL :: ESIMD/tpm_basic.cpp (106 of 273)
16:24:46  PASS: SYCL :: ESIMD/tpm_pointer.cpp (107 of 273)
16:24:47  PASS: SYCL :: ESIMD/tpm_pointer_v2.cpp (108 of 273)
...
16:27:49  Testing Time: 349.95s
16:27:49    Unsupported      :  27
16:27:49    Passed           : 241
16:27:49    Expectedly Failed:   4
16:27:49    Failed           :   1

**Comment by fveselov** [comment]
Results of [PR#2856-No.892](http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Lin/job/LLVM_Test_Suite/892/) job

18:46:40  PASS: SYCL :: ESIMD/tpm_basic.cpp (105 of 273)
...
18:46:43  PASS: SYCL :: ESIMD/tpm_pointer.cpp (107 of 273)
18:46:43  PASS: SYCL :: ESIMD/tpm_pointer_v2.cpp (108 of 273)
...
18:49:47    Unsupported      :  27
18:49:47    Passed           : 241
18:49:47    Expectedly Failed:   4
18:49:47    Failed           :   1

**Comment by vladimirlaz** [comment]
> Results of [PR#2856-No.892](http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Lin/job/LLVM_Test_Suite/892/) job
> 
> 18:46:40 PASS: SYCL :: ESIMD/tpm_basic.cpp (105 of 273)
> ...
> 18:46:43 PASS: SYCL :: ESIMD/tpm_pointer.cpp (107 of 273)
> 18:46:43 PASS: SYCL :: ESIMD/tpm_pointer_v2.cpp (108 of 273)
> ...
> 18:49:47 Unsupported : 27
> 18:49:47 Passed : 241
> 18:49:47 Expectedly Failed: 4
> 18:49:47 Failed : 1

Could you please make sure that al three possible configurations Linux, Linux_CUDA, Windows. If there are failures please share failed test cases

**Comment by fveselov** [comment]
> > Results of [PR#2856-No.892](http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Lin/job/LLVM_Test_Suite/892/) job
> > 18:46:40 PASS: SYCL :: ESIMD/tpm_basic.cpp (105 of 273)
> > ...
> > 18:46:43 PASS: SYCL :: ESIMD/tpm_pointer.cpp (107 of 273)
> > 18:46:43 PASS: SYCL :: ESIMD/tpm_pointer_v2.cpp (108 of 273)
> > ...
> > 18:49:47 Unsupported : 27
> > 18:49:47 Passed : 241
> > 18:49:47 Expectedly Failed: 4
> > 18:49:47 Failed : 1
> 
> Could you please make sure that al three possible configurations Linux, Linux_CUDA, Windows. If there are failures please share failed test cases

For both [Linux_CUDA](http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Lin/job/LLVM_Test_Suite_CUDA/855/console) and [Windows](http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Win/job/Test_Suite/130/console) platforms all three tests fall into 'Unsupported' category

15:48:36  UNSUPPORTED: SYCL :: ESIMD/tpm_basic.cpp (102 of 273)
15:48:36  UNSUPPORTED: SYCL :: ESIMD/tpm_pointer.cpp (103 of 273)
15:48:36  UNSUPPORTED: SYCL :: ESIMD/tpm_pointer_v2.cpp (104 of 273)

**Review by vladimirlaz — DISMISSED** [review]

File: SYCL/ESIMD/tpm_basic.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
+//==---------------- basic_tpm.cpp  - DPC++ ESIMD on-device test
+//------------==//
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_basic.cpp]: please fix formatting

File: SYCL/ESIMD/tpm_basic.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
+// RUN: %clangxx-esimd -fsycl %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_basic.cpp]: please add description what this test is intended to check

File: SYCL/ESIMD/tpm_basic.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  int err_cnt = 0;
+  for (int j = 0; j < VL; ++j)
+    if (output[j] != o[j])
+      err_cnt += 1;
+
+  if (err_cnt > 0) {
+    std::cout << "FAILED.\n";
+    return 1;
+  }
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_basic.cpp]: freeing `output` is missing

File: SYCL/ESIMD/tpm_pointer.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
+// RUN: %clangxx-esimd -fsycl %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_pointer.cpp]: all comments from the previous test apply here

Comment (kbobrovs) [reply SYCL/ESIMD/tpm_pointer.cpp]: also, description should make it clear how the tests are different

File: SYCL/ESIMD/tpm_pointer_v2.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
+// RUN: %clangxx-esimd -fsycl %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_pointer_v2.cpp]: all comments from the previous test apply here

Comment (kbobrovs) [reply SYCL/ESIMD/tpm_pointer_v2.cpp]: all tests seem to have a lot of overlap. Can they be merged into one using templates/parameters, so that code is reused as much as possible?

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: SYCL/ESIMD/tpm_basic.cpp [hunk 6]
```diff
@@ -111,6 +113,8 @@ int main(void) {
     if (output[j] != o[j])
       err_cnt += 1;
 
+  free(output, ctx);
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_basic.cpp]: I suggest to use `sycl::malloc_shared` and `sycl::free` to avoid confusion

File: SYCL/ESIMD/tpm_basic.cpp [hunk 7]
```diff
@@ -9,9 +8,12 @@
 // TODO enable on Windows and Level Zero
 // REQUIRES: linux && gpu && opencl
 // RUN: %clangxx-esimd -fsycl %s -o %t.out
-// RUN: %HOST_RUN_PLACEHOLDER %t.out
 // RUN: %ESIMD_RUN_PLACEHOLDER %t.out
 
+// This test is intended to use Thread Private Memory (TPM) to support
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_basic.cpp]: I don't quite understand, sorry. What is being tested? BE support for TPM-allocated variables?
BTW, there is no 'thread' concept in SYCL or ESIMD - workitem or subgroup. In ESIMD workitem == CM's thread, but please use SYCL terms.

Comment (fveselov) [reply SYCL/ESIMD/tpm_basic.cpp]: TPM is ESIMD/CM term and, as far as I know, it is some sort of special memory (part of SYCL's Private Memory) that is used in ESIMD in case variable can not be allocated on registers. These tests are to check the correctness of implementation of TPM in CM, that is used as ESIMD vector backend. I'm not sure how to write description in only SYCL terms.

Comment (kbobrovs) [reply SYCL/ESIMD/tpm_basic.cpp]: This is ESIMD test, ESIMD obeys SYCL execution and memory model, so in general it might not always be possible to test CM BE using ESIMD. These tests are under SYCL/ESIMD, so they should use SYCL/ESIMD only.

Comment (kbobrovs) [reply SYCL/ESIMD/tpm_basic.cpp]: AFAICS, CM TPM == SYCL private memory. Private is per-workitem memory.

**Comment by fveselov** [comment]
Re-run after merging tests in one: [PR#2856-No.937](http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Lin/job/LLVM_Test_Suite/937/)

02:31:43  PASS: SYCL :: ESIMD/tpm_tests.cpp (105 of 271)

File: SYCL/ESIMD/tpm_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
+// RUN: %clangxx-esimd -fsycl %s -o %t.out
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out 1
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out 2
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out 3
+
+// Since in ESIMD a single WI occupies entire underlying H/W thread, SYCL
+// private memory maps to what's known as 'thread private memory' in CM.
+// This test is intended to use TPM to support implementation in ESIMD
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_tests.cpp]:
```suggestion
// This test is intended to check TPM support implementation in ESIMD
```

File: SYCL/ESIMD/tpm_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+  {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Test>(
+          sycl::range<1>{1}, [=](id<1> i) SYCL_ESIMD_KERNEL {
+            using namespace sycl::INTEL::gpu;
+            simd<int, VL> val(0);
+
+            int x1[SZ];
+            for (int j = 0; j < SZ; ++j) {
```
Comment (kbobrovs) [inline SYCL/ESIMD/tpm_tests.cpp]: can some of this code be outlined into a function, which can be reused on the host side as well?

**Comment by kbobrovs** [comment]
Please remove 'TPM' from the title as well

File: SYCL/ESIMD/private_memory_access.cpp [hunk 10]
```diff
@@ -0,0 +1,179 @@
+//==--------------- private_memory_access.cpp - DPC++ ESIMD on-device test -==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
```
Comment (vladimirlaz) [inline SYCL/ESIMD/private_memory_access.cpp]: Please note that @DenisBakhvalov  is enabling ESIMD tests on windows and level_zero.
Please make sure that you are aligned (I do expect that it should be supported).

**Review by vladimirlaz — CHANGES_REQUESTED** [review]
NIT: Squashing tests into one test case is not perfect solution. It is better to move source to Input folder and have 3 cpp files with tests run lines only (see https://github.com/intel/llvm-test-suite/tree/intel/SYCL/AOT as example)

File: SYCL/ESIMD/private_memory_access.cpp [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/esimd.hpp>
+#include <iostream>
+
+using namespace cl::sycl;
+
+constexpr unsigned VL = 8;
+constexpr unsigned SZ = 800; // big enough to use TPM
+
+ESIMD_INLINE void work(int *o, int case_num, int offx1, int offx2, int offy1,
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]: IMO, it is better not to pollute test cases with unrelated code. DCE on `if`s with compile-time evaluated conditions should help. Suggestions below implement that.

```suggestion
ESIMD_INLINE template<int CASE_NUM> void work(int *o, int offx1, int offx2, int offy1,
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+      x1[j] = (j * (x1[j] + x1[j - 1]) / divisor) - base2;
+  }
+
+  // some work with X2
+  for (int j = 1; j < SZ; ++j) {
+    if ((x2[j] + j) < base2)
+      x2[j] = (divisor * (x2[j] - x2[j - 1]) / j) + base1;
+  }
+
+  if (case_num == 1) {
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
  if constexpr (CASE_NUM == 1) {
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    // some work with Y2
+    for (int j = 1; j < SZ - 1; j += 2) {
+      if ((*(y2[j]) <= *(y2[j + 1]))) {
+        auto temp = y2[j];
+        y2[j] = y2[j + 1];
+        y2[j + 1] = temp;
+      }
+    }
+
+    if (case_num == 2) {
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
    if constexpr (CASE_NUM == 2) {
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+      for (int j = 0; j < SZ; ++j)
+        o[j % VL] += *(y1[j]) - *(y2[j]);
+    } else { // case_num == 3
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
    } else { // case_num == 3
      static_assert(CASE_NUM == 3, "invalid CASE_NUM");
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        if (j % 18 == 0)
+          (*(*(z[j])))++;
+      }
+
+      for (int j = 0; j < SZ; ++j)
+        o[j % VL] += *(*(z[j]));
+    }
+  }
+}
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
template <int CASE_NUM> class KernelID;
template <int CASE_NUM> int test();
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+  auto ctx = q.get_context();
+
+  int *output =
+      static_cast<int *>(sycl::malloc_shared(VL * sizeof(int), dev, ctx));
+  memset(output, 0, VL * sizeof(int));
+
+  int case_num = atoi(argv[1]);
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
  int case_num = CASE_NUM;
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+  int offy1 = 499;
+  int offy2 = 223;
+  int offz = 99;
+  int base1 = 500;
+  int base2 = 100;
+  int divisor = 4;
+
+  {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Test>(sycl::range<1>{1},
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
      cgh.parallel_for<KernelID<CASE_NUM>>(sycl::range<1>{1},
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+                                   [=](id<1> i) SYCL_ESIMD_KERNEL {
+                                     using namespace sycl::INTEL::gpu;
+
+                                     int o[VL] = {0};
+
+                                     work(o, case_num, offx1, offx2, offy1,
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
                                     work<CASE_NUM>(o, offx1, offx2, offy1,
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+                                     block_store<int, VL>(output, val);
+                                   });
+    });
+    e.wait();
+  }
+
+  int o[VL] = {0};
+
+  work(o, case_num, offx1, offx2, offy1, offy2, offz, base1, base2, divisor);
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
  work<CASE_NUM>(o, offx1, offx2, offy1, offy2, offz, base1, base2, divisor);
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+    }
+  }
+}
+
+int main(int argc, char **argv) {
+  if (argc != 2) {
+    std::cout << "Skipped! Specify case number" << std::endl;
+    return 1;
+  }
+
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
  int case_num = atoi(argv[1]);
  switch (case_num) {
  case 0: return test<0>();
  case 1: return test<1>();
  case 2: return test<2>();
  case 3: return test<3>();
  }
  std::cerr << "Invalid case number: " << case_num << "\n";
  return 1;
}
```

File: SYCL/ESIMD/private_memory_access.cpp [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory_access.cpp]:
```suggestion
template <int CASE_NUM> int test() {
  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
```

**Review by vladimirlaz — APPROVED** [review]

File: SYCL/ESIMD/private_memory/pm_access_1.cpp [hunk 22]
```diff
@@ -0,0 +1,12 @@
+//==--------------- pm_access_1.cpp - DPC++ ESIMD on-device test ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
```
Comment (kbobrovs) [inline SYCL/ESIMD/private_memory/pm_access_1.cpp:10]: Please check with @DenisBakhvalov if `REQUIRES: linux` can be removed - can be done as successive patch.

**Review by kbobrovs — APPROVED** [review]
LGTM, but see a comment

File: SYCL/ESIMD/private_memory/pm_access_1.cpp (hunk 22 above)
Comment (DenisBakhvalov) [reply SYCL/ESIMD/private_memory/pm_access_1.cpp:10]: Today, unfortunately, we cannot remove this restriction. We can revisit this ~2 weeks from now.
